File: 26-Remove Duplicates from Sorted Array/RemoveDuplicatesSortedArray.kt
fun main() {
    val nums = intArrayOf(1, 2, 2)
    val result = RemoveDuplicatesSortedArray.removeDuplicates(nums)

    println(result)
}

abstract class RemoveDuplicatesSortedArray {
    companion object {
        fun removeDuplicates(nums: IntArray): Int {
            // First attempt Mar 24, 2024 20:13
            // Runtime 269ms, Memory 43.1MB
            //
            // The first element always considered unique
//            var replaceIndex = 1
//
//            for (index in 1..<nums.size) {
//                // Check if current value NOT matches the previous value
//                if (nums[index - 1] != nums[index]) {
//                    // Replace item with index [replaceIndex] with current value
//                    nums[replaceIndex] = nums[index]
//                    // Increase the [replaceIndex] to execute next item
//                    replaceIndex++
//                }
//            }
//
//            println("Nums: ${nums.toList()}")
//
//            return replaceIndex
            //
            // Second attempt Mar 24, 2024 21:56
            // Runtime 213ms, Memory 39.9MB
            //
            var replaceIndex = 0
            var currentIndex = 1

            while (currentIndex < nums.size) {
                if (nums[currentIndex] == nums[replaceIndex]) {
                    currentIndex++
                    continue
                }

                replaceIndex++
                nums[replaceIndex] = nums[currentIndex]

                currentIndex++
            }

            return replaceIndex + 1
        }
    }
}

File: 26-Remove Duplicates from Sorted Array/remove_duplicates_sorted_array.go
package main

import "fmt"

func main() {
	nums := []int{1}
	fmt.Println("Result: ", removeDuplicates(nums))
}

func removeDuplicates(nums []int) int {
	// First attempt May 30, 2024 20:32
	// Runtime 8ms, Memory 4.47MB
	//
	unique := 1

	for i := 1; i < len(nums); i++ {
		if nums[i] != nums[i-1] {
			nums[unique] = nums[i]
			unique++
		}
	}

	return unique
}
